networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/client_nsx.py

```diff
@@ -0,0 +1,206 @@
+                eventlet.sleep(pause)
+
+            raise Exception(msg, last_err)
+
+        return decorator
+
+
+
+class Client:
+
+    def __init__(self):
+        rate=cfg.CONF.NSXV3.nsxv3_requests_per_second
+        timeout=cfg.CONF.NSXV3.nsxv3_requests_per_second_timeout
+        limit=1
+
+        self._api_scheduler = Scheduler(rate, limit, timeout)
+
+        self._timeout = cfg.CONF.NSXV3.nsxv3_request_timeout
+
+        self._base_path = 'https://{}:{}'.format(
+            cfg.CONF.NSXV3.nsxv3_login_hostname,
+            cfg.CONF.NSXV3.nsxv3_login_port
+        )
+
+        self._login_timestamp = 0
+        self._login_path = "/api/session/create"
+        self._login_data = {
+            "j_username": cfg.CONF.NSXV3.nsxv3_login_user,
+            "j_password": cfg.CONF.NSXV3.nsxv3_login_password
+        }
+
+        self._session = requests.session()
+
+        if cfg.CONF.NSXV3.nsxv3_suppress_ssl_wornings:
+            self._session.verify = False
+            requests.packages.urllib3.disable_warnings()
+
+        self._version = None
+
+    @property
+    def version(self, refresh=False):
+        if not self._version or refresh:
+            resp = self.get(path="/api/v1/node/version")
+            if resp.ok:
+                self._version = resp.json()['product_version']
+        return versionutils.convert_version_to_tuple(self._version)
+
+    def _login(self):
+        LOG.info("Session token - acquiring")
+        now = int(datetime.datetime.now().strftime("%s"))
+        with LockManager.get_lock(self._base_path):
+            if now > self._login_timestamp:
+                resp = requests.post(**self._params(path=self._login_path,
+                                                    data=self._login_data,
+                                                    verify=self._session.verify))
+                if resp.status_code != requests.codes.ok:
+                    resp.raise_for_status()
+
+                self._session.headers["Cookie"] = \
+                    resp.headers.get("Set-Cookie")
+                self._session.headers["X-XSRF-TOKEN"] = \
+                    resp.headers.get("X-XSRF-TOKEN")
+                self._session.headers["Accept"] = "application/json"
+                self._session.headers["Content-Type"] = "application/json"
+
+                self._login_timestamp = \
+                    int(datetime.datetime.now().strftime("%s"))
+
+        try:
+            # Refresh version after login
+            self.version(refresh=True)
+        except Exception:
+            pass
+        LOG.info("Session token - acquired, connected to NSX-T {}".format(self._version))
+
+    def _params(self, **kwargs):
+        kwargs["timeout"] = self._timeout
+        kwargs["url"] = "{}{}".format(self._base_path, kwargs["path"])
+        del kwargs["path"]
+        if self._login_path not in kwargs["url"] and "data" in kwargs:
+            kwargs["data"] = json.dumps(kwargs["data"])
```

Comment:
I'd like to understand why we use the manual `json.dumps()` here and not the `json` argument requests provides. Is it because that would also set the `application/json` content-type?

Reply:
The main reason was to login, as it uses the post but not json. But will make it another way. `json` arguments makes more sense. Thanks